File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/client/VerifyingKeyId.java
package org.hyperledger.iroha.android.client;

import java.util.Objects;
import org.hyperledger.iroha.android.model.zk.VerifyingKeyBackendTag;

/** Exact identifier returned by the Torii verifying-key registry id projection. */
public final class VerifyingKeyId {
  private final String backend;
  private final String name;

  VerifyingKeyId(final String backend, final String name) {
    this.backend = Objects.requireNonNull(backend, "backend");
    this.name = Objects.requireNonNull(name, "name");
  }

  /** Exact verifier-registry backend label. */
  public String backend() {
    return backend;
  }

  /** Exact verifying-key registry name. */
  public String name() {
    return name;
  }

  /** Low-level proof engine bound to this exact verifier-registry label. */
  public VerifyingKeyBackendTag engine() {
    final VerifyingKeyBackendTag engine =
        VerifyingKeyBackendTag.verifierBackendRegistryTagV1(backend);
    if (engine == null) {
      throw new IllegalStateException("unsupported verifier-registry backend " + backend);
    }
    return engine;
  }

  @Override
  public boolean equals(final Object other) {
    if (this == other) {
      return true;
    }
    if (!(other instanceof VerifyingKeyId)) {
      return false;
    }
    final VerifyingKeyId that = (VerifyingKeyId) other;
    return backend.equals(that.backend) && name.equals(that.name);
  }

  @Override
  public int hashCode() {
    return Objects.hash(backend, name);
  }

  @Override
  public String toString() {
    return backend + ":" + name;
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/client/VerifyingKeyId.kt
package org.hyperledger.iroha.sdk.client

import org.hyperledger.iroha.sdk.core.model.zk.VerifyingKeyBackendTag

/** Exact identifier returned by Torii's verifying-key registry id projection. */
class VerifyingKeyId internal constructor(
    /** Exact verifier-registry backend label. */
    @JvmField val backend: String,
    /** Exact verifying-key registry name. */
    @JvmField val name: String,
) {
    /** Low-level proof engine bound to this exact verifier-registry label. */
    fun engine(): VerifyingKeyBackendTag =
        checkNotNull(VerifyingKeyBackendTag.verifierBackendRegistryTagV1(backend)) {
            "unsupported verifier-registry backend $backend"
        }

    override fun equals(other: Any?): Boolean =
        this === other ||
            other is VerifyingKeyId && backend == other.backend && name == other.name

    override fun hashCode(): Int = 31 * backend.hashCode() + name.hashCode()

    override fun toString(): String = "$backend:$name"
}
